### ISSUE TITLE
A way to use custom parser for a language

### ISSUE THREAD

**Issue by MunifTanjim** [nvim-treesitter/nvim-treesitter#2241]
**Is your feature request related to a problem? Please describe.**

I want to use the https://github.com/MunifTanjim/tree-sitter-lua parser for Lua.

From the discussion at https://github.com/neovim/neovim/pull/15260, it looks like `query_file_ignore` is not a good solution for using a custom parser.

So I'm trying to install it with a different name.

I've tried this:

```lua
  local parser_config = require("nvim-treesitter.parsers").get_parser_configs()

  parser_config.lua_neo = {
    install_info = {
      url = "https://github.com/MunifTanjim/tree-sitter-lua",
      files = { "src/parser.c", "src/scanner.c" },
      location = "tree-sitter-lua_neo/lua",
      revision = "a5fc8b29c5c74fe016f4da71badb9476277b5146",
    },
    filetype = "lua",
    maintainers = { "@MunifTanjim" },
  }
```

And `:TSInstall lua_neo` works.

But then I get this error when it is used:

```
Error detected while processing FileType Autocommands for "*":
E5108: Error executing lua Failed to load parser: uv_dlsym: /home/shaikat/.local/share/nvim/plugged/nvim-treesitter/parser/lua_neo.so: undefined symbol: tree_sitter_lua_neo
stack traceback:
        [C]: in function '_ts_add_language'
        ...local/share/nvim/runtime/lua/vim/treesitter/language.lua:33: in function 'require_language'
        ...shaikat/.local/share/nvim/runtime/lua/vim/treesitter.lua:57: in function '_create_parser'
        ...shaikat/.local/share/nvim/runtime/lua/vim/treesitter.lua:112: in function 'get_parser'
        ...lugged/nvim-treesitter/lua/nvim-treesitter/highlight.lua:107: in function 'attach'
        .../plugged/nvim-treesitter/lua/nvim-treesitter/configs.lua:458: in function 'attach_module'
        .../plugged/nvim-treesitter/lua/nvim-treesitter/configs.lua:481: in function 'reattach_module'
        [string ":lua"]:1: in main chunk
```

So, it looks like I need to change the name in the parser itself. But this parser is indeed for `lua` language, not `lua_neo`. So changing it in parser doesn't make sense.

**Describe the solution you'd like**

I'd like a simple way to satisfy this use-case. I'm willing to do it myself and create a PR if somebody points me to the right direction.

**Describe alternatives you've considered**

`@clason` [suggested](https://github.com/neovim/neovim/pull/15260#issuecomment-1009827033):

> In this case, however, if your Lua parser is indeed better than the current fork in nvim-treesitter, we might wish to switch -- but that is again something handled by nvim-treesitter, so I'd suggest opening an issue there about switching (where we may also want to consider `@tjdevries`' parser, so it would be good to include a comparison of the three).

But I don't wanna pursue that route. Reason being:
- It would cause trouble for many plugins that ships with queries for the default lua parser that ships with `nvim-treesitter`.
- Just because I think my parser is better, doesn't mean it is.

So I think providing a simple way to use custom parser with `nvim-treesitter` would be a better solution.

**Additional context**
- https://github.com/neovim/neovim/issues/15244
- https://github.com/neovim/neovim/pull/15260

**Comment by theHamsta** [comment]
In my opinion, there should be a better way to tie the queries to a specific parser. There's a similar problem when a parser implementation in nvim-treesitter is switched or nvim-treesitter queries collide with built-in ones

There's two possibilities on how you use an alternative parser:
- you only want to use the new one -> replace the old lua parser `lua` as the language name is baked into the binary and is determined by this line: https://github.com/MunifTanjim/tree-sitter-lua/blob/a5fc8b29c5c74fe016f4da71badb9476277b5146/lua/grammar.js#L27
- you want to use both parsers and possibly switch between them on runtime: this is only possible when you replace https://github.com/MunifTanjim/tree-sitter-lua/blob/a5fc8b29c5c74fe016f4da71badb9476277b5146/lua/grammar.js#L27 by some other name, e.g. `lua_neo`. Then, you configuration should work

**Comment by MunifTanjim** [comment]
> you only want to use the new one -> replace the old lua parser lua

Yes, I only want to use the new one. But only replacing the `lua` parser won't work. Because all the plugins with `queries/lua` will have queries for old parser. And we don't have any way to specify which query files should be used (we only have `set_query` which accepts a string, not very friendly).


> Tree-sitter will try to load tree_sitter_XXX(), and in C land there should be only one symbol with the same name loaded at runtime. uv_dlsym fails because it tries to find tree_sitter_neo_lua but can't find the symbol.

Yes, I understand that issue. If I rename that to `lua_neo` or something else in `MunifTanjim/tree-sitter-lua` it would work without any changes to `nvim-treesitter` or `neovim`. But that would be a workaround, since that parser is actually meant to replace the `Azganoth/tree-sitter-lua`.

> In my opinion, there should be a better way to tie the queries to a specific parser.

Although it is not very common, there can be alternative parsers for the same language. And providing a better way to tie the queries to a specific parser would really help me in this case.

Any suggestion about how the API for "a better way to tie the queries to a specific parser" might look like?

---

(I've been rebasing https://github.com/neovim/neovim/pull/15260 with latest master and compiling neovim from that branch for the last 6 months. Which isn't much pain for me, I just run a script for doing that. But I don't wanna keep that branch hanging forever if that's not gonna be merged. That's why I created this issue to discuss a better alternative.)

**Comment by theHamsta** [comment]
@MunifTanjim maybe the one could just load the base queries from the directory on which the parser is installed. Philosophy: you bring the parser, you're responsible for the queries or just have a way that you can install your parser as `lua_neo` with `lua_neo` queries without actually having to rename the `tree_sitter_lua` in your grammar.js/parser.c

Still think that just having different parser names would be a better solution:
- nvim-treesitter could provide queries for multiple parser implementation
- could switch parsers on runtime
- no confusion which queries is meant for which parsers
Down sides:
- language injection would need to have a way to dispatch between the different parsers

**Comment by MunifTanjim** [comment]
Thanks!

I went with this one:

> maybe the one could just load the base queries from the directory on which the parser is installed. Philosophy: you bring the parser, you're responsible for the queries

https://github.com/MunifTanjim/nvim-treesitter-lua/blob/5b4e0050da9aa913f65d1050b21766929047de8a/lua/nvim-treesitter-lua/init.lua#L24-L30

I think this can be closed now.

**Comment by clason** [comment]
No, please don't close it -- just because _your_ concrete problem is solved, the general issue about using different parsers, and how to have better control over which queries are loaded, is still valuable to discuss.

(And your concrete problem may still be solved differently anyway: by just switching nvim-treesitter's Lua parser, which is still under discussion.)

**Comment by theHamsta** [comment]
> maybe the one could just load the base queries from the directory on which the parser is installed. Philosophy: you bring the parser, you're responsible for the queries

With this I wasn't referring, to just doing that in your instance but as a way to bind a parser to it's queries

**Comment by clason** [comment]
I think this needs to be a coordinated discussion with upstream, but I'm not sure whether it makes sense to open an issue there yet. Can this be separated into independent features (parser vs. query loading)?
